src/Microsoft.OpenApi/Services/OpenApiUrlTreeNode.cs

```diff
@@ -0,0 +1,225 @@
+        /// </summary>
+        /// <param name="doc">The OpenAPI document.</param>
+        /// <param name="label">Name tag for labelling the <see cref="OpenApiUrlTreeNode"/> nodes in the directory structure.</param>
+        /// <returns>The root node of the created <see cref="OpenApiUrlTreeNode"/> directory structure.</returns>
+        public static OpenApiUrlTreeNode Create(OpenApiDocument doc, string label)
+        {
+            Utils.CheckArgumentNull(doc, nameof(doc));
+            Utils.CheckArgumentNullOrEmpty(label, nameof(label));
+
+            OpenApiUrlTreeNode root = new OpenApiUrlTreeNode("/");
+
+            var paths = doc.Paths;
+            if (paths != null)
+            {
+                foreach (var path in paths)
+                {
+                    root.Attach(path: path.Key,
+                                pathItem: path.Value,
+                                label: label);
+                }
+            }
+
+            return root;
+        }
+
+        /// <summary>
+        /// Retrieves the paths from an OpenAPI document and appends the items to an <see cref="OpenApiUrlTreeNode"/> node.
+        /// </summary>
+        /// <param name="doc">The OpenAPI document.</param>
+        /// <param name="label">Name tag for labelling related <see cref="OpenApiUrlTreeNode"/> nodes in the directory structure.</param>
+        public void Attach(OpenApiDocument doc, string label)
+        {
+            Utils.CheckArgumentNull(doc, nameof(doc));
+            Utils.CheckArgumentNullOrEmpty(label, nameof(label));
+
+            var paths = doc?.Paths;
+            if (paths != null)
+            {
+                foreach (var path in paths)
+                {
+                    Attach(path: path.Key,
+                           pathItem: path.Value,
+                           label: label);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Appends a path and the PathItem to an <see cref="OpenApiUrlTreeNode"/> node.
+        /// </summary>
+        /// <param name="path">An OpenAPI path.</param>
+        /// <param name="pathItem">Path Item object that describes the operations available on an OpenAPI path.</param>
+        /// <param name="label">A name tag for labelling the <see cref="OpenApiUrlTreeNode"/> node.</param>
+        /// <returns>An <see cref="OpenApiUrlTreeNode"/> node describing an OpenAPI path.</returns>
+        public OpenApiUrlTreeNode Attach(string path,
+                                         OpenApiPathItem pathItem,
+                                         string label)
+        {
+            Utils.CheckArgumentNullOrEmpty(label, nameof(label));
+
+            if (path.StartsWith("/"))
+            {
+                // Remove leading slash
+                path = path.Substring(1);
+            }
+
+            var segments = path.Split('/');
```

Comment:
remove empty entries?

Reply:
Well, we can. But in either case, it would still work okay, since we're checking for null or empty strings either way here: https://github.com/irvinesunday/OpenAPI.NET/blob/7d36d697524aeb43369477d3b297996381256045/src/Microsoft.OpenApi/Services/OpenApiUrlTreeNode.cs#L168

![image](https://user-images.githubusercontent.com/40403681/115597302-cb795580-a2e1-11eb-8813-ceb91d41540e.png)